artemis/modules/dangling_dns_detector.py

```diff
@@ -0,0 +1,201 @@
+                if answer.rdtype == record_type:
+                    dangling = False
+                    break
+
+        return dangling
+
+    def check_cname(self, domain: str, result: list[dict[str, Any]]) -> None:
+        try:
+            answers = dns.resolver.resolve(domain, rdatatype.CNAME, raise_on_no_answer=False)
+            if answers.rrset is not None:
+                for record in answers:
+                    dangling = self.__check_cname(record)
+                    if dangling:
+                        result.append(
+                            {
+                                "domain": domain,
+                                "record": rdatatype.CNAME,
+                                "message": (
+                                    "The defined domain has CNAME record configured but the CNAME does not resolve."
+                                ),
+                            }
+                        )
+        except (dns.resolver.NXDOMAIN, dns.resolver.NoNameservers, dns.resolver.Timeout):
+            pass
+
+    def check_soa(
+        self, domain: str, ns_name: dns.name.Name, server_ip: ipaddress.IPv4Address, result: list[dict[str, Any]]
+    ) -> None:
+        name = ns_name
+        if not name.is_absolute():
+            name = name.concatenate(dns.name.root)
+
+        reply = direct_dns_query(name, rdatatype.SOA, server_ip)
+        if not reply:
+            result.append({"domain": domain, "record": rdatatype.NS, "message": "Target NS query failed."})
+            return None
+
+    def __check_ns(
+        self,
+        domain: str,
+        qname: dns.name.Name,
+        record: Rdata,
+        result: list[dict[str, Any]],
+    ) -> bool | None:
+        if not hasattr(record, "rdtype") or record.rdtype != rdatatype.NS:
+            return None
+
+        ns_target = record.target.to_text()  # type: ignore[attr-defined]
+
+        correct_responses = 0
+        for record_type in [rdatatype.A, rdatatype.AAAA]:
+            response = edns_query(ns_target, record_type)
+
+            if not response:
+                continue
+
+            for answer in response:
+                if record_type == rdatatype.A:
+                    correct_responses += 1
+                    self.check_soa(domain, qname, ipaddress.IPv4Address(answer.address), result)  # type: ignore[attr-defined]
+                elif record_type == rdatatype.AAAA:
+                    correct_responses += 1
+
+        return correct_responses == 0
+
+    def check_ns(self, domain: str, result: list[dict[str, Any]]) -> None:
+        # It may generate false positives
+        try:
+            answers = dns.resolver.resolve(domain, rdatatype.NS, raise_on_no_answer=False)
+            if answers.rrset is not None:
+                for record in answers:
+                    dangling = self.__check_ns(domain, answers.qname, record, result)
+                    if dangling:
+                        result.append(
```

Comment:
I think returning a list is more Pythonic than modifying a list passed as reference